misc/mdns/src/service.rs

```diff
@@ -398,44 +454,66 @@ impl<'a> MdnsResponse<'a> {
-    pub fn discovered_peers<'b>(&'b self) -> impl Iterator<Item = MdnsPeer<'b>> {
+    pub fn discovered_peers<'b>(&'b self) -> impl Iterator<Item = MdnsPeer> + 'b {
         let packet = &self.packet;
         self.packet.answers.iter().filter_map(move |record| {
-            if record.name.to_string().as_bytes() != SERVICE_NAME {
-                return None;
-            }
-
-            let record_value = match record.data {
-                RData::PTR(record) => record.0.to_string(),
-                _ => return None,
-            };
+            let name = record.name.to_string();
+            
+            if name.as_bytes() == SERVICE_NAME {
 
-            let peer_name = {
+                let record_value = match record.data {
+                    RData::PTR(record) => record.0.to_string(),
+                    _ => return None,
+                };
+            
                 let mut iter = record_value.splitn(2, |c| c == '.');
-                let name = match iter.next() {
+                let peer_name = match iter.next() {
                     Some(n) => n.to_owned(),
                     None => return None,
                 };
-                if iter.next().map(|v| v.as_bytes()) != Some(SERVICE_NAME) {
-                    return None;
+
+                let decoded = match iter.next().map(|v| v.as_bytes()) {
+                    Some(SERVICE_NAME) => data_encoding::BASE32_DNSCURVE.decode(peer_name.as_bytes()),
+                    _ => return None
+                }.map(|bytes| PeerId::from_bytes(bytes));
+
+                if let Ok(Ok(peer_id)) = decoded {
+                    return Some(MdnsPeer::parse(
+                        peer_id,
+                        record.ttl,
+                        packet,
+                        record_value,
+                    ))
                 }
-                name
-            };
-
-            let peer_id = match data_encoding::BASE32_DNSCURVE.decode(peer_name.as_bytes()) {
-                Ok(bytes) => match PeerId::from_bytes(bytes) {
-                    Ok(id) => id,
-                    Err(_) => return None,
-                },
-                Err(_) => return None,
-            };
-
-            Some(MdnsPeer {
-                packet,
-                record_value,
-                peer_id,
-                ttl: record.ttl,
-            })
+            } else if name.as_bytes().ends_with(LEGACY_SERVICE_NAME_JS) {
+                let mut iter = name.splitn(2, |c| c == '.');
+                if let Ok(peer_id) = match iter.next() {
+                    Some(n) => n.parse::<PeerId>(),
+                    None => return None,
+                } {
+                    if let RData::SRV(srv) = record.data {
+                        // FIXME: assuming ipv4
+                        let addr = match format!(
+                            // FIXME: dns4 is probably incorrect
+                            "/dns4/{}/tcp/{}/p2p/{}",
```

Comment:
two issues here:
 - js-libp2p-mdns sends over `os.hostname()` - which isn't necessarily accessible over LAN
 - `dns4` is probably _wrong_ for this use case, `dnsaddrs` might be more appropriate but it is unclear how these are supposed to interact from the spec and we can't parse it yet -- see #967 --

Dialing of this addrs appears to fail anyway.

Reply:
`js-libp2p-mdns` also sends an A record for the hostname.

Reply:
Yeah, I noticed that, but it was a lot more pain to get into the code base, so it took a lot longer - but it's in there now: https://github.com/libp2p/rust-libp2p/pull/968/files#diff-030eefcb7ad763f8dd6732ebef838e0eR490